### DISCUSSION TITLE
Download Limiter Not Working?

### DISCUSSION THREAD

**Discussion by mepherion** [tnodir/fort#215] [Q&A]
In Fort Firewall, I set the Application Group download speed limit to 125829 kb/s (around 125 Mbps), and set Brave browser to that Application Group.

![Fort Firewall App Group Setting](https://github.com/tnodir/fort/assets/20154933/0cd9689e-1fb7-4420-9bed-2257712b9e56)
![Fort Firewall App Setting](https://github.com/tnodir/fort/assets/20154933/8c45af3a-1763-44a1-9716-74f12f166316)

When I run the speedtest, I only get ~55 Mbps.
![Fort Firewall Speedtest](https://github.com/tnodir/fort/assets/20154933/51fd9103-fe6f-4785-93ef-bad06965223b)

-----

However, using win-shaper with the same setting, it works perfectly on speedtest which limits to ~125 Mbps
![win-shapper setting](https://github.com/tnodir/fort/assets/20154933/cbf6a60f-e6f8-42f1-8d83-044d1a5af676)
![win-shapper](https://github.com/tnodir/fort/assets/20154933/db144f38-96f1-4816-8852-1daa3e07fd56)

Am I doing something wrong?

**Comment by tnodir** [comment]
I've found a typo in the code. Let me fix it..

**Reply by tnodir** [reply]
@mepherion Please check the [v3.11.6](https://github.com/tnodir/fort/releases/tag/v3.11.6).

**Reply by mepherion** [reply]
No change with the new version, unless I need to restart my computer after applying the update.

![image2](https://github.com/tnodir/fort/assets/20154933/73112ef2-67b8-439a-86c4-2e7e9a12c969)

**Reply by mepherion** [reply]
@tnodir It seems like the download/upload speed limit kind of works with numbers lower than 50 Mbps, but not exact. Anything higher, it doesn't work at all.

5 Mbps Limit (observed around 10 Mbps):
![5 Mbps](https://github.com/tnodir/fort/assets/20154933/e2a5093d-bd11-4073-99f3-5bef9e7c6506)

50 Mbps Limit (observed around 40 Mbps):
![50 Mbps](https://github.com/tnodir/fort/assets/20154933/1ba23331-96b0-4585-a2b9-f4aaebbcc38b)

100 Mbps Limit (observed around 40 Mbps):
![100 Mbps](https://github.com/tnodir/fort/assets/20154933/fd462a57-302d-4ed5-b8c2-9181a2d85a85)

Unlimited:
![Unlimited](https://github.com/tnodir/fort/assets/20154933/d9bfc82a-d253-4c11-9c48-c752702d737a)

**Comment by tnodir** [comment]
@mepherion Thank you for testing this feature.

Fort’s implementation differs from Win-Shape: packets processed in transport layer (to support Windows 7/8), not Ethernet frame.

I tested only Download part with 5Mb/s limit, when implemented it. But Upload does not work correctly with such low speeds too, as I’ve tested yesterday.

I have to rework some parts of speed limiter to fix the bug..

**Reply by tnodir** [reply]
@mepherion Please check the [v3.11.7](https://github.com/tnodir/fort/releases/tag/v3.11.7).

**Reply by mepherion** [reply]
v3.11.7 causes a BSOD with error `BAD_POOL_CALLER`

| Dump File           | Bug Check String | Bug Check Code | Parameter 1 | Parameter 2 | Parameter 3 | Parameter 4 | Caused By Driver | Caused By Address | File Description      | Product Name                         | Company               | Crash Address       | Stack Address 1     | Stack Address 2 | Stack Address 3 |
| ------------------- | ---------------- | -------------- | ----------- | ----------- | ----------- | ----------- | ---------------- | ----------------- | --------------------- | ------------------------------------ | --------------------- | ------------------- | ------------------- | --------------- | --------------- |
| 022024-31765-01.dmp | BAD_POOL_CALLER  | 0x000000c2     | 0x0000000d  | 0xd2302c04  | 0x04080000  | 0xa356c317  | NETIO.SYS        | NETIO.SYS+639c    | Network I/O Subsystem | Microsoft® Windows® Operating System | Microsoft Corporation | ntoskrnl.exe+17ef3c | ntoskrnl.exe+2ad060 | NETIO.SYS+639c  | tcpip.sys+c080f |

![image](https://github.com/tnodir/fort/assets/20154933/e5ad57dd-7687-4b63-9cb8-2b3b074f55df)

**Reply by tnodir** [reply]
Please send the Minidump file to nodir.temir@gmail.com

**Reply by tnodir** [reply]
@mepherion Please check the [v3.11.9](https://github.com/tnodir/fort/releases/tag/v3.11.9).

**Reply by mepherion** [reply]
Same BSOD error with v3.11.9, sent minidump via email

**Reply by tnodir** [reply]
How can I reproduce the BSOD? What speed limits to set? What program to use? Which site to open, etc?

**Reply by mepherion** [reply]
I put the limiter to the preset of 50 Mbps and ran the speedtest on https://www.speedtest.net

**Reply by tnodir** [reply]
@mepherion Does it BSOD with [v3.12.0-test1](https://github.com/tnodir/fort/releases/tag/v3.12.0-test1) too?

**Reply by tnodir** [reply]
@mepherion We can not reproduce the BSOD with 50 Mbps limits on speedtest.

Do you have any Antivirus or other Firewall?

**Reply by tnodir** [reply]
@mepherion Can you please also provide info about hardware?
- network adapter model
- tests are via Ethernet cable or WiFi

Also do you use VPN or IPSec?

**Comment by ghost** [comment]
I don't get a BSOD, but in my case I use LAN on this laptop, it is a _Intel(R) Ethernet Connection I219-LM_, when I limit downloads speeds doesn't cause issues, the problem is if I limit the upload sleep.

In my case, I create a group TEMP, and set it to 10Mbps upload limit, then select Edge to use TEMP, if I start using Brave everything is alright, but the moment I try to open a page in Edge and try to navigate, it will affect everything, and then it will say there is a connection problem, and my whole adapter crashed. It will start saying DNS problems, then appear as disconnected and then **unidentified network**, and I lose connection, and the adapter will not work anymore and the only way to bring it back is to restart the computer.

I tested **v3.12.0-test1**, and it is the same. 
I don't know if WIFI would cause the same. Once I told you that my ethernet froze and I had to manually shutdown my computer, seems like I was right and this was the reason, I remember it didn't make my computer's Ethernet crash so fast though, now it is like immediately.

Since I have to restart on every test, I don't really tested it much and often to record a video explaining it, it happens every time the way I typed it, also I don't really limit speed that much and less upload, just in rare times, but this was my experience. Download speed limiter seems like it doesn't cause issues with my Ethernet port or the whole network adapters, but just explaining what happens here.

I am on Win11 22H2, I was thinking of going 23H2 so might test it again if I ever upgrade.

**Reply by tnodir** [reply]
@Emi-Emi-Emi Thanks for checking it. Can you please test some other speed limiter, e.g. NetLimiter? Will it cause same issues with network adapter?

**Reply by ghost** [reply]
Send my comment with the wrong account lol but yeah.

**Comment by tnodir** [comment]
I'm thinking about reworking the Speed Limiter's code in the Driver as in win-shaper.

But it will work only on Windows 8+.

Do anyone need the Speed Limits on Windows 7?

**Reply by tnodir** [reply]
No, it's not possible to process packets per App in the Mac Frame layer.
